### PR TITLE
[MRG] use glob with check_niimg

### LINKED ISSUE #763: Have the niimgs checking function optionally convert globbing
SHould we expand it? It would be more convenient for me but I don't know if it's YAGNI or not.

### PR DESCRIPTION
**Pull request by aabadie**
Fixes #763

### CODE REVIEW

File: nilearn/_utils/niimg_conversions.py [hunk 1]
```diff
@@ -176,6 +181,15 @@ def check_niimg(niimg, ensure_ndim=None, atleast_4d=False, dtype=None,
         data will be converted to int32 if dtype is discrete and float32 if it
         is continuous.
 
+    return_iterator: boolean, optional
+        Returns an iterator on the content of the niimg file input
+
+    widlcards: boolean, optional
+        Use wildcards to list matching input images. If multiple files
+        matches and return iterator is set, the returned list of objects
```
Comment (bthirion) [inline nilearn/_utils/niimg_conversions.py]: 'match'
return_iterator

File: nilearn/_utils/niimg_conversions.py [hunk 2]
```diff
@@ -176,6 +181,15 @@ def check_niimg(niimg, ensure_ndim=None, atleast_4d=False, dtype=None,
...
+        is sorted (ascending order), otherwise the first object in the list is
+        returned.
+
```
Comment (bthirion) [inline nilearn/_utils/niimg_conversions.py]: I find this ('return the first object of the list') a strange default behavior. maybe I did not understand.

**Comment by bthirion** [comment]
Looks rather good, but documentation should be added to illustrate the behavior.

File: nilearn/_utils/niimg_conversions.py [hunk 3]
```diff
@@ -13,8 +14,11 @@
 from .niimg import _safe_get_data, load_niimg
 from .compat import _basestring, izip
 
-from nilearn._utils.exceptions import DimensionError
+from .exceptions import DimensionError
+from .. import EXPAND_PATH_WILDCARDS
 
+if EXPAND_PATH_WILDCARDS:
+    import glob
```
Comment (lesteve) [inline nilearn/_utils/niimg_conversions.py]: Not really necessary to import glob conditionally.

Comment (aabadie) [inline nilearn/_utils/niimg_conversions.py]: Ok ! I removed it in 810643a

**Comment by lesteve** [comment]
First of all, there is a problem that `nilearn.EXPAND_PATH_WILDCARDS` and `nilearn._utils.niimg_conversions.EXPAND_PATH_WILDCARDS` are two different variables, so setting one has no effect on the other. As an illustration:

```python
from nilearn._utils import niimg_conversions
niimg_conversions.check_niimg('this_file_does_not_exist')  # error with please set global constants EXPAND_PATH_WILDCARDS to False

import nilearn
nilearn.EXPAND_PATH_WILDCARDS = False
niimg_conversions.check_niimg('this_file_does_not_exist')  # still the same error as previously
```

Also I am a bit confused about EXPAND_PATH_WILDCARDS functionality. Why don't we want to always use `EXPAND_PATH_WILDCARDS = True` ? 

Is it because we worry about files which contains the following characters: `*`, `[` or `?` ?

File: nilearn/_utils/niimg_conversions.py [hunk 4]
```diff
@@ -193,6 +207,22 @@ def check_niimg(niimg, ensure_ndim=None, atleast_4d=False, dtype=None,
...
+        # processing filenames matching globbing expression
+        if len(filenames) > 1:
+            niimg = filenames  # iterable case
+        elif len(filenames) == 1:
+            # Only one file matching => loading it as is
+            niimg = filenames[0]
+        else:
+            # No files matching the glob expression, warn the user
+            raise ValueError("No files matching the entered niimg "
+                             "expression : %s.\n You may have left wildcards "
```
Comment (lesteve) [inline nilearn/_utils/niimg_conversions.py]: Missing % niimg or something to put the user input in the error message.

**Comment by GaelVaroquaux** [comment]
> First of all, there is a problem that nilearn.EXPAND_PATH_WILDCARDS and
> nilearn._utils.niimg_conversions.EXPAND_PATH_WILDCARDS

Indeed, we need to do 'import nilearn as ni' in
nilearn._utils.niimg_conversions and then access EXPAND_PATH_WILDCARDS as
"ni.EXPAND_PATH_WILDCARDS"

**Comment by aabadie** [comment]
> Indeed, we need to do 'import nilearn as ni'

I pushed this change in commit 01b479b.

**Comment by GaelVaroquaux** [comment]
> I pushed this change in commit 01b479b.

You need a test that fails before this commit and works after, that way
we are convinced that it works.

**Comment by lesteve** [comment]
> Indeed, we need to do 'import nilearn as ni' in nilearn._utils.niimg_conversions and then access EXPAND_PATH_WILDCARDS as "ni.EXPAND_PATH_WILDCARDS"

Note that I couldn't figure out a way to do the same thing with a relative import. Not sure how likely this is to create a problem in the real world.

> You need a test that fails before this commit and works after, that way we are convinced that it works.

:+1:

**Comment by aabadie** [comment]
> You need a test that fails before this commit and works after, that way we are convinced that it works.
+1
I pushed 3115ba7 which contains a tests covering more use cases related to globbing. I'm waiting for the tests to fail and will push the fix later.

**Comment by GaelVaroquaux** [comment]
> Note that I couldn't figure out a way to do the same thing with a relative
> import.

Good point.

> Not sure how likely this is to create a problem in the real world.

Let's not loose sleep on this and move on.

**Comment by GaelVaroquaux** [comment]
>     I pushed 3115ba7 which contains a tests covering more use cases related to
>     globbing. I'm waiting for the tests to fail and will push the fix later.

Test driven development! Yey

**Comment by aabadie** [comment]
Ok, so tests are now completely failing... (but not for the initial reason). @GaelVaroquaux maybe you prefer to validate the tests first or is it ok if I push my solution now ?

File: nilearn/_utils/niimg_conversions.py [hunk 5]
```diff
@@ -193,6 +207,28 @@ def check_niimg(niimg, ensure_ndim=None, atleast_4d=False, dtype=None,
...
+        if (wildcards and
+            ni.EXPAND_PATH_WILDCARDS):
+            filenames = sorted(glob.glob(niimg))  # Ascending sorting
+
+            # processing filenames matching globbing expression
+            if len(filenames) > 1:
+                niimg = filenames  # iterable case
+            elif len(filenames) == 1:
+                # Only one file matching => loading it as is
+                niimg = filenames[0]
```
Comment (AlexandreAbraham) [inline nilearn/_utils/niimg_conversions.py]: Same remark as before: in nilearn, a list of one file may not be an error nor a single file. This case is hard to handle. Maybe at least a warning would be enough ?

File: nilearn/_utils/niimg_conversions.py [hunk 6]
```diff
@@ -193,6 +207,28 @@ def check_niimg(niimg, ensure_ndim=None, atleast_4d=False, dtype=None,
...
+            else:
+                # No files matching the glob expression, warn the user
+                message = ("No files matching the entered niimg expression: "
+                       "'%s'.\n You may have left wildcards usage activated: "
+                       "please set the global constant 'EXPAND_PATH_WILDCARDS' "
+                       "to False or use option 'wildcards=False' to deactivate "
+                       "this behavior.") % niimg
+                raise ValueError(message)
```
Comment (AlexandreAbraham) [inline nilearn/_utils/niimg_conversions.py]: Just a guess: can't we raise a warning and automatically retry by setting wildcards to False?

**Comment by AlexandreAbraham** [comment]
Apart from some usability questions, :+1:

File: nilearn/_utils/niimg_conversions.py [hunk 7]
```diff
@@ -176,6 +179,17 @@ def check_niimg(niimg, ensure_ndim=None, atleast_4d=False, dtype=None,
         data will be converted to int32 if dtype is discrete and float32 if it
         is continuous.
 
+    return_iterator: boolean, optional
+        Returns an iterator on the content of the niimg file input
+
+    wildcards: boolean, optional
```
Comment (KamalakerDadi) [inline nilearn/_utils/niimg_conversions.py]: It would be nice if you refer some where giving reference to small code snippet to here. If you think its good, it is located in ```doc/manipulating_visualizing/manipulating_images.rst``` file. It would be more clear enough.

Comment (aabadie) [inline nilearn/_utils/niimg_conversions.py]: Added online documentation in  243f789

File: nilearn/tests/test_niimg_conversions.py [hunk 8]
```diff
@@ -191,6 +193,97 @@ def test_check_niimg():
...
 
 
+def test_check_niimg_wildcards():
+    tmp_dir = tempfile.tempdir + os.sep
+    nofile_path = "/tmp/nofile"
+    nofile_path_wildcards = "/tmp/no*file"
+    wildcards_msg = ("No files matching the entered niimg expression: "
+                     "'%s'.\n You may have left wildcards usage "
+                     "activated: please set the global constant "
+                     "'EXPAND_PATH_WILDCARDS' to False to deactivate "
```
Comment (lesteve) [inline nilearn/tests/test_niimg_conversions.py]: You should probably be more explicit and say `nilearn.EXPAND_PATH_WILDCARDS`

File: nilearn/_utils/niimg_conversions.py [hunk 9]
```diff
@@ -193,6 +207,30 @@ def check_niimg(niimg, ensure_ndim=None, atleast_4d=False, dtype=None,
...
+                niimg = filenames  # iterable case
+            elif len(filenames) == 1:
+                # Only one file matching => loading it as is
+                niimg = filenames[0]
+            elif glob.has_magic(niimg):
+                # No files matching the glob expression, warn the user
+                message = ("No files matching the entered niimg expression: "
+                            "'%s'.\n You may have left wildcards usage "
+                           "activated: please set the global constant "
+                           "'EXPAND_PATH_WILDCARDS' to False to deactivate "
```
Comment (lesteve) [inline nilearn/_utils/niimg_conversions.py]: > You should probably be more explicit and say nilearn.EXPAND_PATH_WILDCARDS

I should have done the original comment here.

File: nilearn/tests/test_niimg_conversions.py [hunk 10]
```diff
@@ -191,6 +193,97 @@ def test_check_niimg():
...
+    with testing.write_tmp_imgs(data_img, data_img,
+                                create_files=True,
+                                use_wildcards=True) as globs:
+        assert_array_equal(_utils.check_niimg(tmp_dir + globs).get_data(),
+                           img_4d.get_data())
+
+    #######
+    # Test when global variable is set to False => no globbing allowed
+    ni.EXPAND_PATH_WILDCARDS = False
+    assert_equal(ni.EXPAND_PATH_WILDCARDS, False)
```
Comment (lesteve) [inline nilearn/tests/test_niimg_conversions.py]: This is really overkill isn't it?

File: nilearn/tests/test_niimg_conversions.py [hunk 11]
```diff
@@ -191,6 +193,97 @@ def test_check_niimg():
...
+                        _utils.check_niimg, nofile_path, wildcards=False)
+
+    # Testing with an exact filename
+    with testing.write_tmp_imgs(data_img, create_files=True) as filename:
+        assert_array_equal(_utils.check_niimg(filename).get_data(),
+                           data_img.get_data())
+
+    # Reverting to default behavior
+    ni.EXPAND_PATH_WILDCARDS = True
+    assert_equal(ni.EXPAND_PATH_WILDCARDS, True)
```
Comment (lesteve) [inline nilearn/tests/test_niimg_conversions.py]: Same comment, a bit overkill...

File: nilearn/_utils/niimg_conversions.py (hunk 9 above)
Comment (KamalakerDadi) [inline nilearn/_utils/niimg_conversions.py]: Will it be not easy to mention only about ```wildcards=False``` ?

Comment (aabadie) [inline nilearn/_utils/niimg_conversions.py]: As this message could be raised in the user console and wildcards=False is used internally (check_niimg should not be called directly by users), I would stick to the global constant in the message.

This makes me think that the wildcards option is useless... and should be removed from the function signature.

Comment (aabadie) [inline nilearn/_utils/niimg_conversions.py]: > You should probably be more explicit and say nilearn.EXPAND_PATH_WILDCARDS

ok

File: nilearn/_utils/niimg_conversions.py (hunk 5 above)
Comment (lesteve) [inline nilearn/_utils/niimg_conversions.py]: Maybe we should do: `elif [niimg] == filenames` to differentiate the case where niimg is just a filename vs the one where the glob only matches one file.  In the latter I think it should be considered as a list with a single file.

**Comment by aabadie** [comment]
@lesteve @KamalakerDadi your recent comments have been adressed ;)

File: nilearn/_utils/niimg_conversions.py (hunk 9 above)
Comment (KamalakerDadi) [inline nilearn/_utils/niimg_conversions.py]: > I would stick to the global constant in the message.

Its upto you but my point is user should not look to change global constants to set to False. That's my intuition. Correct me if I am wrong.

**Comment by aabadie** [comment]
@KamalakerDadi, @GaelVaroquaux @lesteve @AlexandreAbraham, is this ok to merge ?

**Comment by lesteve** [comment]
Maybe an additional test for the edge case where a real glob (i.e. containing `*`, `[`, or `?`) matches only one file?

**Comment by aabadie** [comment]
> Maybe an additional test for the edge case where a real glob (i.e. containing *, [, or ?) matches only one file?

It's already there I think : see line 235 in test_niimg_conversions.py

**Comment by lesteve** [comment]
> It's already there I think : see line 235 in test_niimg_conversions.py

Good point I missed that. Looking at your tests a little bit more I think we should be testing 3d data too.

If you only have test.nii which is a 3d image in the current working directory, you expect.

```python
 # 3d image
check_niimg('test.nii')
 # 4d image with last dimension 1 i.e. it is equivalent to check_niimg(['test.nii'])
check_niimg('test*.nii')
```

**Comment by aabadie** [comment]
Thanks for your comment @lesteve : I pushed 2 commits that should address those use cases.

**Comment by lesteve** [comment]
Merging this one, thanks!

Still a few regexp rather than glob remaining but I'll fix them up in master directly.

**Comment by aabadie** [comment]
Thanks, this will make my day !
